README.md
# amb-utils

Small FreeDOS utilities for Mateusz Viste's .AMB ([Ancient Machine Book](http://amb.osdn.io/)) e-book format.

[![wtf](wtfpl-badge-1.png)](http://www.wtfpl.net/)

------------------------------------------------------

## AMBRARY
(AMB library, geddit?)  
Written in QuickBASIC 4.5 for FreeDOS 1.3 RC3, 2021. Don't bother trying it in DOSBOX or DosEMU - Their ancient version of COMMAND.COM does not have the DIR switches needed.

Ambrary is a simple launcher menu for your .AMB files. Requires the AMB reader and at least one AMB file to be installed. Well, it wouldn't be much use otherwise, would it?

The command AMBRARY, with no further parameters, will scan your C drive for AMB files

![amb1](ambrary1.png)

and will display up to 128 of them to read in the AMB reader program.

![amb2](ambrary2.png)

### Usage:

If you keep your .AMB files elsewhere, you can start Ambrary with a parameter indicating that. e.g.
~~~~
AMBRARY D:
~~~~
E: and F: will also be accepted. If you give more than one, the highest one will be used.

You can also increase the number of books listed. 
~~~~
AMBRARY 256
~~~~
Accepted values are 256, 512 and 1024. Again, the highest number given is used. If AMB ever takes off to the extent that you need to access more ebooks than that, you will probably need a more sophisticated launcher program.

Parameters may be given in any order. AMBRARY D: 512 and AMBRARY 512 D: work exactly the same.

AMBRARY /D: or AMBRARY --1024 or AMBRARY -F: /256 will all work, since these traditional parameter indicators are ignored.

AMBRARY /? or AMBRARY --HELP provides further onscreen help for newer parameters.

--------------------------------------------------------

## AMBTITLE

A helper program for AMBRARY. Written in QuickBASIC 4.5 for FreeDOS 1.3 RC3,

The command AMBTITLE, with no further parameters, will scan your C drive and extract the title of each book found. This information will be stored in a file that AMBRARY 0.2 or higher can read. AMBRARY will then display the title rather than the filename. from v0.2, entries are automatically sorted by title.

![ambrary3](ambrary3.png)

If you keep your .AMB files elsewhere, you can start AMBTITLE with a parameter indicating that, just like AMBRARY.

You only need to run this utility when you have added new AMB books to your collection or remove them. The whole process is a bit slow, so I have not incorporated this into AMBRARY itself.

AMBTITLE requires DELTREE, the AMBPACK utility and at least one AMB file to be installed.

--------------------------------------------------------
## Scripts

Some Bash scripts that I worked out to create .amb files. Yes, although FreeDOS is likely to be the main platform for .amb ebooks, I prefer to create mine on Linux. Sue me.

These are for my own purposes. They are full of hard-coded paths and don't pretend to be universal solutions. But perhaps they can help someone work out their own workflow.

--------------------------------------------------------

## AMB e-books

The AMB directory contains a few e-books I have created in AMB format, mainly for my own purposes and so that I can get a handle on this format. Right now, I can't say how far this library will grow.

I will only include AMB files in languages I actually know, but if you'd like to see your AMB file in Tagalog, Serbo-Croatian or Guarani listed here, [email them to me](mailto:clasqm@gmail.com) and we'll see what we can do.



CHANGES
AMBRARY

v0.1 Initial release
V0.2 - Will now read AMBTITLE-generated database
     - If database is present, will default to using that. Either an 
       AMBTITLE or a regular AMBRARY rescan can be forced with parameters.
v0.3 - Now works with sorted AMBTITLE database.
     - Will warn and exit if v0.1 database is detected.

AMBTITLE

v0.1 - Initial release
v0.2 - auto-sorts database: fields are reversed from v0.1



scripts/dosambcreate.sh
#!/bin/sh

# Creates FreeDOS-ready .amb ebooks (codepage 437) on linux once
# all the .ama files have been prepared.

# Should work on all West European languages. 437 is a "least common
# denominator" in DOS 8-bit code pages, and ebooks made with this
# should remain readable to users using CP850 and CP858 (Don't use box-
# drawing characters). Includes a premade 437 to UTF8 conversion map
# for use on *NIX systems.

# Made for my own purposes - filled to the brim with hardcoded paths
# but feel free to adapt it for your own purposes.

cd ~/Unzip/etexts/workspace
mkdir final
cp title ./final/
cp ../unicode.map ./final/
for i in *.ama; do utf8tocp 437 $i > final/$i; done
cd final
echo filename?
read
ambpack c . /home/michel/Dropbox/devuan/git/amb-utils/AMB/$REPLY
cd /home/michel/Dropbox/devuan/git/amb-utils/AMB/
amb $REPLY
read -r -p "Delete temporary files? [Y/n] " input
 case $input in
    [yY][eE][sS]|[yY])
cd ~/Unzip/etexts/workspace
rm -f ./final/*
rmdir ./final
 ;;
    [nN][oO]|[nN])
 echo "OK, then"
       ;;
    *)
 echo "Invalid input..."
 exit 1
 ;;
esac



ambtitle/0.2/AMBTITLE.BAS
'AMBTITLE v0.2

'Written in QuickBASIC 4.5 for FreeDOS, 2021.

'Helper program for AMBRARY

'Scans C: for AMB files and extracts the title for each one.
'AMBRARY v0.2 and later will be able to pick up this title and
'display it instead of the path and filename

'Requires the AMBPACK utility and at least one AMB file to be installed.

DECLARE SUB HelpMe ()
DECLARE SUB EndThis ()
DECLARE SUB AmbScan ()
DECLARE SUB FillArrays ()
DECLARE SUB SaveFile ()

'$STATIC

'Global string variables
COMMON SHARED tempdir$, tempfile$, ambdrive$
'Global arrays
COMMON SHARED filename$(), title$()
'Global numeric variables
COMMON SHARED amblimit%
IF INSTR(COMMAND$, "HELP") OR INSTR(COMMAND$, "/H") OR INSTR(COMMAND$, "/?") THEN HelpMe
ambdrive$ = "C:"
IF INSTR(COMMAND$, "D:") THEN ambdrive$ = "D:"
IF INSTR(COMMAND$, "E:") THEN ambdrive$ = "E:"
IF INSTR(COMMAND$, "F:") THEN ambdrive$ = "F:"
tempdir$ = ENVIRON$("TEMP")
tempfile$ = tempdir$ + "\ambrary.lst"
AmbScan     'any way to speed this up?
FillArrays  'feed temp file into array
SaveFile    'save results to a file
EndThis

SUB AmbScan
    scanbat% = FREEFILE
    OPEN tempdir$ + "\rescan.bat" FOR OUTPUT AS #scanbat%
    PRINT #scanbat%, "@echo off"
    PRINT #scanbat%, "cls"
    PRINT #scanbat%, "echo."
    PRINT #scanbat%, "echo           *******************************"
    PRINT #scanbat%, "echo           ******** AMBTITLE scan ********"
    PRINT #scanbat%, "echo           *******************************"
    PRINT #scanbat%, "echo           *                             *"
    PRINT #scanbat%, "echo           *  Scanning " + ambdrive$ + " for AMB files  *"
    PRINT #scanbat%, "echo           *  This could take a while:   *"
    PRINT #scanbat%, "echo           *  DOS can be slow sometimes. *"
    PRINT #scanbat%, "echo           *                             *"
    PRINT #scanbat%, "echo           *******************************"
    PRINT #scanbat%, "echo."
    PRINT #scanbat%, "dir /S /B /ON /-P " + ambdrive$ + "\*.amb > " + tempfile$
    CLOSE #scanbat%
    SHELL tempdir$ + "\rescan.bat"
    KILL tempdir$ + "\rescan.bat"
END SUB

SUB EndThis
    CLOSE
    END
END SUB

SUB FillArrays
    'remove leftovers
    SHELL "deltree /y " + tempdir$ + "\amb > NUL"
    'get number of items
    tmp1% = FREEFILE
    OPEN tempfile$ FOR INPUT AS #tmp1%
        f = 0
        DO
            LINE INPUT #tmp1%, a$
            f = f + 1
        LOOP UNTIL EOF(tmp1%)
    CLOSE #tmp1%
    'DIM arrays
    DIM filename$(f)
    DIM title$(f)
    tmp1% = FREEFILE
    OPEN tempfile$ FOR INPUT AS #tmp1%
        f = 1
        DO UNTIL EOF(tmp1%)
            LINE INPUT #tmp1%, a$
            IF a$ <> "" THEN
                filename$(f) = a$
                PRINT "Found " + a$
                PRINT "Extracting ..."
                SHELL "ambpack u " + a$ + " " + tempdir$ + "\amb > NUL"
                tmp2% = FREEFILE
                OPEN tempdir$ + "\amb\title" FOR INPUT AS #tmp2%
                LINE INPUT #tmp2%, c$
                title$(f) = LEFT$(c$, LEN(c$) - 1)'strip trailing LF
                PRINT "Title: " + c$
                CLOSE #tmp2%
                SHELL "deltree /y " + tempdir$ + "\amb > NUL"
                f = f + 1
            END IF
        LOOP
    IF filename$(1) = "" THEN
        PRINT
        PRINT "No AMB files found on drive " + ambdrive$
        EndThis
    END IF
    CLOSE #tmp1%
    KILL tempfile$
END SUB

SUB HelpMe
    PRINT "AMBTITLE v0.2"
    PRINT "A helper program for AMBRARY."
    PRINT
    PRINT "The command AMBTITLE, with no further parameters, will scan your"
    PRINT "C drive and extract the title of each book found. This information"
    PRINT "will be stored in a file that AMBRARY 0.2 or higher can read."
    PRINT
    PRINT "If you keep your .AMB files elsewhere, you can start AMBTITLE with a"
    PRINT "parameter indicating that. e.g. AMBTITLE D:. Drives E: and F: will"
    PRINT "also be accepted. If you give more than one, the highest one is used."
    PRINT
    PRINT "You only need to run this utility when you have added new AMB books"
    PRINT "to your collection or remove them."
    PRINT
    PRINT "AMBTITLE requires DELTREE, SORT, the AMBPACK utility and at least"
    PRINT "one AMB file to be installed."
    PRINT
    EndThis
END SUB

SUB SaveFile
    PRINT "Writing results to file ..."
    PRINT
    tmp1% = FREEFILE
    OPEN "c:\fdos\doc\ambtitle.tmp" FOR OUTPUT AS #tmp1%
    FOR f = 1 TO UBOUND(filename$)
        PRINT #tmp1%, title$(f) + CHR$(9) + filename$(f)
    NEXT f
    CLOSE #tmp1%
    PRINT "Sorting file ..."
    SHELL "sort /A c:\fdos\doc\ambtitle.tmp > c:\fdos\doc\ambtitle.db"
    KILL "c:\fdos\doc\ambtitle.tmp"
END SUB



ambrary/0.3/AMBRARY.BAS
'AMBRARY v0.3
'(AMB library, geddit?)

'Written in QuickBASIC 4.5 for FreeDOS, 2021.
'Released under the Dowhatthefuckyouwant license.

'Scans C: for AMB files and allow user to select one for reading.

'Requires the AMB reader and at least one AMB file to be installed.
'Well, it wouldn't be much use otherwise, would it?

'This version handles up to 124 AMB files by default. Should be enough for now,
'but otherwise just change the amblimit% variable below and recompile.

DECLARE SUB Display ()
DECLARE SUB HelpMe ()
DECLARE SUB endthis ()
DECLARE SUB AmbScan ()
DECLARE SUB ReadDB (filename$)
DECLARE SUB FillArray ()
DECLARE FUNCTION ScrollBox% (lim%, high%, wide%, LIN%, COL%)
DECLARE FUNCTION DbaseExists% (filename$)

'$STATIC

'Global string variables
COMMON SHARED tempdir$, tempfile$, ambdrive$, dbfile$
'Global arrays
COMMON SHARED filename$(), title$()
'Global numeric variables
COMMON SHARED amblimit%
IF INSTR(COMMAND$, "HELP") OR INSTR(COMMAND$, "/H") OR INSTR(COMMAND$, "/?") THEN HelpMe
amblimit% = 128
IF INSTR(COMMAND$, "256") THEN amblimit% = 256
IF INSTR(COMMAND$, "512") THEN amblimit% = 512
IF INSTR(COMMAND$, "1024") THEN amblimit% = 1024
DIM filename$(amblimit%)
DIM title$(amblimit%)
ambdrive$ = "C:"
IF INSTR(COMMAND$, "D:") THEN ambdrive$ = "D:"
IF INSTR(COMMAND$, "E:") THEN ambdrive$ = "E:"
IF INSTR(COMMAND$, "F:") THEN ambdrive$ = "F:"
tempdir$ = ENVIRON$("TEMP")
tempfile$ = tempdir$ + "\ambrary.lst"
dbfile$ = "C:\FDOS\DOC\AMBTITLE.DB"
IF INSTR(COMMAND$, "AMBTITLE") THEN SHELL "ambtitle " + ambdrive$
IF INSTR(COMMAND$, "RESCAN") THEN
    AmbScan     'any way to speed this up?
    FillArray   'feed temp file into array
    Display
    endthis
END IF
IF DbaseExists%(dbfile$) = 1 THEN
    ReadDB (dbfile$)
ELSE
    AmbScan
    FillArray   'feed temp file into array
END IF
Display
endthis

SUB AmbScan
    scanbat% = FREEFILE
    OPEN tempdir$ + "\rescan.bat" FOR OUTPUT AS #scanbat%
    PRINT #scanbat%, "@echo off"
    PRINT #scanbat%, "cls"
    PRINT #scanbat%, "echo."
    PRINT #scanbat%, "echo           *******************************"
    PRINT #scanbat%, "echo           ********* AMBRARY scan ********"
    PRINT #scanbat%, "echo           *******************************"
    PRINT #scanbat%, "echo           *                             *"
    PRINT #scanbat%, "echo           *  Scanning " + ambdrive$ + " for AMB files  *"
    PRINT #scanbat%, "echo           *  This could take a while:   *"
    PRINT #scanbat%, "echo           *  DOS can be slow sometimes. *"
    PRINT #scanbat%, "echo           *                             *"
    PRINT #scanbat%, "echo           *******************************"
    PRINT #scanbat%, "echo."
    PRINT #scanbat%, "dir /S /B /ON /-P " + ambdrive$ + "\*.amb > " + tempfile$
    CLOSE #scanbat%
    SHELL tempdir$ + "\rescan.bat"
    KILL tempdir$ + "\rescan.bat"
END SUB

FUNCTION DbaseExists% (filename$)
    tmp1% = FREEFILE
    OPEN filename$ FOR BINARY AS #tmp1%
    IF LOF(tmp1%) <> 0 THEN DbaseExists% = 1
    CLOSE #tmp1%
END FUNCTION

SUB Display
    SCREEN 0
    CLS
    DO
        IF choice% <> 0 THEN
            SHELL "amb " + filename$(choice%)
            choice% = 0
        ELSE
            choice% = ScrollBox%(amblimit% + 1, 23, 76, 1, 1)
        END IF
    LOOP
END SUB

SUB endthis
    CLOSE
    END
END SUB

SUB FillArray
    tmp1% = FREEFILE
    filename$(0) = "*********** EXIT ***********"
    OPEN tempfile$ FOR INPUT AS #tmp1%
    FOR f = 1 TO amblimit%
        IF NOT EOF(tmp1%) THEN
            LINE INPUT #tmp1%, a$
            IF a$ <> "" THEN
                filename$(f) = a$
            END IF
        END IF
    NEXT f
    IF filename$(1) = "" THEN
        PRINT
        PRINT "No AMB files found on drive " + ambdrive$
        endthis
    END IF
    CLOSE #tmp1%
    KILL tempfile$
END SUB

SUB HelpMe
    PRINT "AMBRARY v0.3"
    PRINT "A simple launcher menu for your .AMB (Ancient Machine Book) files."
    PRINT
    PRINT "USAGE: AMBRARY  [D: | E: | F:] [256 | 512 | 1024] [RESCAN] [AMBTITLE]"
    PRINT
    PRINT "The command AMBRARY, with no further parameters, will load a data"
    PRINT "file produced by AMBTITLE. If this is not found, AMBRARY will scan"
    PRINT "the C drive for AMB files and will display up to 128 of them to"
    PRINT "select for reading in the AMB reader program."
    PRINT
    PRINT "If you keep your .AMB files elsewhere, you can start Ambrary with a"
    PRINT "a parameter indicating that. e.g. AMBRARY D: . E: and F: will also be"
    PRINT "accepted. If you give more than one, the highest one will be used."
    PRINT
    PRINT "You can also increase the number of books listed. Accepted values are"
    PRINT "256, 512 and 1024. Again, the highest number given is used."
    PRINT
    PRINT "You can force a normal rescan with the RESCAN parameter, even if the"
    PRINT "AMBTITLE database is detected. Or you can force a slower AMBTITLE rescan"
    PRINT "with the AMBTITLE parameter (AMBTITLE must be installed)."
    PRINT
    PRINT "AMBRARY D: 512 and AMBRARY 512 D: work exactly the same way."
    PRINT
    PRINT "AMBRARY /D: or AMBRARY --1024 or AMBRARY -F: /256 will all work,"
    PRINT "since these traditional parameter indicators are ignored."
    endthis
END SUB

SUB ReadDB (filename$)
    tmp1% = FREEFILE
    OPEN filename$ FOR INPUT AS #tmp1%
        entries% = 0
        DO
            LINE INPUT #tmp1%, dummy$
            entries% = entries% + 1
        LOOP UNTIL EOF(tmp1%)
        CLOSE #tmp1%
        IF entries% > amblimit% THEN entries% = amblimit%
    filename$(0) = "*********** EXIT ***********"
    title$(0) = "*********** EXIT ***********"
    OPEN filename$ FOR INPUT AS #tmp1%
    FOR f = 1 TO entries%
        IF NOT EOF(tmp1%) THEN
            LINE INPUT #tmp1%, a$
            'changed the order of fields so I could sort them
            'will have to keep this in for a few versions.
            IF MID$(a$, 2, 2) = ":\" THEN
                PRINT "Old-style (v.01) database detected."
                PRINT "Please rescan with AMBTITLE v0.2 or later"
                endthis
            END IF
            IF a$ <> "" THEN
                FOR n = 1 TO LEN(a$)
                    IF MID$(a$, n, 1) = CHR$(9) THEN MIDPOINT% = n
                NEXT n
                title$(f) = LEFT$(a$, MIDPOINT% - 1)
                filename$(f) = MID$(a$, MIDPOINT% + 1)
            END IF
        END IF
    NEXT f
    IF filename$(1) = "" THEN
        PRINT
        PRINT "No AMB files found on drive " + ambdrive$
        endthis
    END IF
    CLOSE #tmp1%
END SUB

FUNCTION ScrollBox% (lim%, high%, wide%, LIN%, COL%)
REM ===========================================================================
REM Author: Kurt Kuzba
REM I made it a FUNCTION, as opposed to a SUB, because it is also
REM useful for menus, and you can always just disregard the return
REM value.
REM
REM Adapted by Michel Clasquin-Johnson for his own nefarious purposes.
REM ===========================================================================
    REM lim%= number of lines to display
    COLOR 7, 1: CLS : sel% = 0: top% = 0: ex% = lim%
    COLOR 1, 7: LOCATE 1, 1, 0: s$ = STRING$(wide%, " ")
    FOR t% = 0 TO high% + 1
        LOCATE LIN% + t%, COL%
        PRINT STRING$(wide% + 4, CHR$(177));
    NEXT
    LIN% = LIN% + 1: COL% = COL% + 2
    WHILE lim% = ex%
        COLOR 9, 1: LOCATE LIN% - 1, COL%
        PRINT MID$(STR$(sel% + 1), 2) + STRING$(3, CHR$(177))
        FOR t% = 0 TO (high% - 1)
            LOCATE LIN% + t%, COL%: COLOR 1, 7
            IF t% + top% = sel% THEN COLOR 7, 1
            IF title$(top% + t%) <> "" THEN
                PRINT LEFT$(title$(top% + t%) + s$, wide%);
            ELSE
                PRINT LEFT$(filename$(top% + t%) + s$, wide%);
            END IF
        NEXT: p% = 0: p$ = " HPIQGO" + CHR$(13)
        WHILE p% < 2: p% = INSTR(p$, RIGHT$(INKEY$, 1)): WEND
        SELECT CASE p%
            CASE IS = 2   'CURSOR UP
                IF sel% > 0 THEN sel% = sel% - 1
                IF sel% < top% THEN top% = sel%
            CASE IS = 3   'CURSOR DOWN
                IF sel% < (lim% - 1) THEN sel% = sel% + 1
                IF sel% > (top% + (high% - 2)) THEN top% = sel% - high% + 1
            CASE IS = 4   'PAGE UP
                top% = top% - high%: sel% = sel% - high%
                IF top% < 0 THEN top% = 0
                IF sel% < 0 THEN sel% = 0
            CASE IS = 5   'PAGE DOWN
                top% = top% + high%: sel% = sel% + high%
                IF top% >= lim% - high% THEN top% = lim% - high%
                IF top% < 0 THEN top% = 0
                IF sel% >= lim% THEN sel% = lim% - 1
            CASE IS = 6   'HOME
                top% = 0: sel% = 0
            CASE IS = 7   'END
                sel% = lim% - 1: top% = sel% - high% + 1
                IF top% < 0 THEN top% = 0
            CASE IS = 8   'ENTER
                lim% = sel%
        END SELECT
    WEND
    IF sel% = 0 THEN endthis 'only way out of an endless loop. Yes, I know,
    ScrollBox% = sel%        'bad form to put it here and not in the main module.
END FUNCTION                 'So sue me.

